---
title: "Integrated analysis"
author: "Suzanne Scott"
date: "10/28/2021"
output:
  html_document:
    keep_md: true
    df_print: paged
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    code_folding: hide
params:
  include_mouse: false
---

After we have imported and QC'ed the data, we proceed with analysis using Seurat.

```{r include=FALSE}
knitr::opts_chunk$set(fig.width = 7, fig.height=5, dev = c('pdf', 'png'))

library(Seurat)
library(dplyr)
library(patchwork)
source("workflow.R")

if (params$include_mouse) {
  data.obj <- import_with_mouse()
} else {
  data.obj <- import_without_mouse()
}

save.file <- glue::glue("../out/Seurat/integrated/mouse-{params$include_mouse}")
dir.create(dirname(save.file), recursive=TRUE)
```

## Nomalization, highly variable genes

Next, we normalise the data, using a "LogNormalize' method.  Also, analysis in Seurat focuses on a subset of genes that exhibit high cell-to-cell variation in the dataset (highly expressed in some cells, lowly expressed in others).  The next step is to identify these genes.

```{r normalize, include=FALSE}

for (d in names(data.obj)) {
  all.genes <- rownames(data.obj[[d]])
  data.obj[[d]] <- NormalizeData(data.obj[[d]], normalization.method = "LogNormalize", scale.factor = 10000)
  
  data.obj[[d]] <- FindVariableFeatures(data.obj[[d]], selection.method = "vst", nfeatures = 2000)
}
```


Plot these features for each dataset

```{r variableGenes, fig.width=25, fig.height=15, warning=FALSE}
plotlist = list()
for (sample in names(data.obj)) {
  top5 <- head(VariableFeatures(data.obj[[sample]]), 5)
  plotlist[[sample]] <- VariableFeaturePlot(data.obj[[sample]])
  plotlist[[sample]] <- LabelPoints(plot=plotlist[[sample]], points=top5, repel=TRUE, xnudge=0, ynudge=0) +
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::ggtitle(sample)
}
cowplot::plot_grid(plotlist = plotlist)
```

## Integration

At this point, we can combine the datasets by finding integration anchors.  These are cross-dataset pairs of cells that are in a matched biological state ('anchors') that can be used to correct for both technical differences between datasets (i.e. batch effect correction) and to perform comparative analysis across experimental conditions.

```{r integrate, include=FALSE}
features <- SelectIntegrationFeatures(object.list = data.obj)
anchors <- FindIntegrationAnchors(object.list = data.obj, anchor.features = features, dims=1:30)
cells <- IntegrateData(anchorset = anchors, dims=1:30)
DefaultAssay(cells) <- "integrated"
```


```{r annotateIntegrated}
get_dataset <- function(data_obj, barcode) {
  barcode_check <- stringr::str_split(barcode, "_")[[1]][2]
  return(names(data_obj)[as.integer(barcode_check)])
}

get_sample_type <- function(data_obj, barcode) {
  return(
    stringr::str_replace(
      get_dataset(data_obj, barcode), 
      "\\d+$", 
      ""
    )
  )
}



# add column to split data on 
cells[['dataset_ind']] <- purrr::map_chr(colnames(cells), 
                                             ~stringr::str_split(., "_")[[1]][2])

cells[['dataset_name']] <- purrr::map_chr(colnames(cells), 
                                             ~get_dataset(data.obj, .))
cells[['dataset_type']] <- purrr::map_chr(colnames(cells),
                                            ~get_sample_type(data.obj, .)
                                         )
cells[['LK03C_REDHV']] <- ifelse(cells[['dataset_type']] == "LK03C_REDHV",
                                 "LK03C_REDHV", NA)

cells[['LK03V_REDHC']] <- ifelse(cells[['dataset_type']] == "LK03V_REDHC",
                                 "LK03V_REDHC", NA)
```


## Dimensionality reduction - PCA

We next run the standard Seurat workflow on the integrated data. First we scale the data, then we run PCA on the data.


```{r PCA}

# Run the standard workflow for visualization and clustering
cells <- ScaleData(cells, verbose = FALSE)
cells <- RunPCA(cells, npcs = 30, verbose = FALSE)
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA.

First we look at the top genes associated with the first five principal components

```{r PCAdims}
print(cells[['pca']], dims=1:5, nfeatures=5)
```


## Clustering

Seurat uses graph-based clustering to cluster cells with similar properties.  It first constructs a KNN grpah based on the euclidean distance in PCA space, and then it refines the edge weights betweeen any two cells based on the shared overlap in their local neighbourhoods (Jaccard similiarity).

```{r cluster, include=FALSE}
cells <- FindNeighbors(cells, reduction = "pca", dims = 1:30)
cells <- FindClusters(cells, resolution = 0.5)
```


```{r cellsMeta}
cells.meta <- tibble(cells@meta.data)
```


There were `r length(unique(cells.meta$seurat_clusters))`, clusters and each cluster has the following number of cells in each cluster:

```{r clusterCellCounts}
cells.meta %>% 
  group_by(dataset_name, seurat_clusters) %>% 
  summarise(count = n()) %>% 
  tidyr::pivot_wider(names_from="seurat_clusters", values_from="count")
```

### UMAP

We do a nonlinear PCA (UMAP) to visualise the clusters:

```{r clusterDimPlot, include=FALSE}
cells <- RunUMAP(cells, reduction = "pca", dims = 1:30)
p1 <- DimPlot(cells, reduction="umap", label=TRUE, repel=TRUE)
p2 <- DimPlot(cells, reduction="umap", label=FALSE, group.by="dataset_name")
```


```{r}
p1 + p2
```

Here I've colored points instead by which dataset the cell comes from:

```{r clusterDimPlot2}


DimPlot(cells, reduction="umap", split.by="dataset_name", ncol=5)

```

And here I've combined the replicates of the same sample types:

```{r clusterDimPlot3}
DimPlot(cells, reduction="umap", split.by="dataset_type", ncol=3)
```


## Differently expressed features in each cluster - all cells

We first look at the genes that define the clusters, regardless of the sample from which the cells in that cluster came.  The following table shows the top five genes that are the most different in each cluster compared the the rest of the cells.

In other words (the words of [Jared Andrews](https://www.biostars.org/p/409790/)), this analysis will:

> find markers differentially expressed in each identity group by comparing it to all of the others - you don't have to manually define anything. Note that markers may bleed over between closely-related groups - they are not forced to be specific to only one group. 

```{r clusterMarkers}
cell.markers <- FindAllMarkers(cells, only.pos=FALSE, min.pct=0.25, logfc.threshold = 0.25)


saveRDS(cells, file=glue::glue("{save.file}.rds"))
write.table(cell.markers, file=glue::glue("{save.file}_find-all-markers.tsv"), sep="\t")
```


```{r topCellMarkers}
cell.markers %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)
```


We can also plot a heatmap of the top genes for each cluster

```{r clusterHeatmap, fig.height=15, fig.width=15}
top10 <- cell.markers %>% 
  group_by(cluster) %>% 
  top_n(n=10, wt = avg_log2FC)

DoHeatmap(cells, features=top10$gene)
```

## Conserved cell type markers

We next look at the genes that are [conserved between two groups](https://satijalab.org/seurat/reference/findconservedmarkers):

> find markers that are conserved between two groups - this can be useful if you want to find markers that are conserved between a treated and untreated condition for a specific cell type or group of cells. It means they are differentially expressed compared to other groups, but have similar expression between the two groups you're actually comparing.

I do this for each cluster, and compare between the sample types (hepatocytes, low engraftment, high engraftment, vector).  So the genes identified here are conserved in each cluster, irrespective of the source of the cells.

```{r conservedMarkers}

# get number of clusters

DefaultAssay(cells) <- "RNA"
cells.conservedMarkers <- tibble()


for (i in unique(cells.meta$seurat_clusters)) {
  
  # check there are more than three cells in all sample types for  this cluster
  doAnalysis <- cells.meta %>% 
  group_by(dataset_type, seurat_clusters) %>% 
  summarise(count = n()) %>% 
    filter(seurat_clusters == i) %>% 
    pull(count) 
  
  doAnalysis <- all(doAnalysis > 3)
  
  if (doAnalysis) {
  conservedMarkers <- FindConservedMarkers(cells, ident.1 = i, grouping.var = "dataset_type", verbose = FALSE) 
  
  conservedMarkers <- conservedMarkers %>% 
    mutate(gene = rownames(conservedMarkers)) %>% 
    mutate(cluster = i)
  
  cells.conservedMarkers <- bind_rows(cells.conservedMarkers, conservedMarkers)
  write.table(cells.conservedMarkers[[i]], file=glue::glue("{save.file}_conservedClust{i}.tsv"), sep="\t")    
  
  }
  else {
    print(glue::glue(
      "skipping cluster {i} since not all groups have more than three cells"
    ))
  }

}

```
These are the genes within each cluster that are differentially expressed between the conditions, ranked by minimum p value:

```{r conservedMarkersShow}
cells.conservedMarkers %>% 
  select(cluster, gene, minimump_p_val, max_pval, low_p_val:high_p_val_adj) %>% 
  arrange(minimump_p_val)
```



### Genes of interest - Zonation

We can also look at the genes of interest with regards to zonation.  These genes are:

```{r zonationGenesCombined, fig.height = 15, fig.width=15}


genes_of_interest <- function(genes) {
  if (!params$include_mouse) {
    genes <- stringr::str_replace(genes, "Homo-w-AAV-", "")
  }
  p1 <- VlnPlot(cells, features = genes, log=TRUE)
  p2 <- FeaturePlot(cells, features = genes)
  print(glue::glue("A violin plot of the expression of these genes in each cluster:"))
  print(p1)
  print(glue::glue("How these genes are expressed:"))
  print(p2)
}


gene_of_interest <- function(gene) {
  if (!params$include_mouse) {
    gene <- stringr::str_replace(gene, "Homo-w-AAV-", "")
  }  
  detected <- any(stringr::str_detect(rownames(cells), gene))
  if (detected) {
    p1 <- VlnPlot(cells, features = gene, log=TRUE, split.by="dataset_type")
    
    print(p1)
    plotlist <- FeaturePlot(cells, features = gene, split.by="dataset_type", 
                            keep.scale="all", order=TRUE, combine=FALSE)
    
    for (p in seq(length(plotlist))){
      plotlist[[p]] <- plotlist[[p]] +
        ggplot2::theme_classic() + 
        ggplot2::theme(legend.position="none") +
        ggplot2::ggtitle(NULL) +
        ggplot2::labs(x="UMAP_1", y="UMAP_2") 
    }
    
    legend <- cowplot::get_legend(
      plotlist[[3]] + ggplot2::theme(legend.position="bottom")
    )
    plotlist[[p+1]] <- legend
    
    p2 <- cowplot::plot_grid(plotlist=plotlist)
    
    title <- cowplot::ggdraw() + 
      cowplot::draw_label(
        gene,
        fontface = 'bold',
        x = 0,
        hjust = 0
      ) +
      ggplot2::theme(
        # add margin on the left of the drawing canvas,
        # so title is aligned with left edge of first plot
        plot.margin = ggplot2::margin(0, 0, 0, 7)
      )
    
    print(cowplot::plot_grid(title, p2, ncol=1, rel_heights=c(0.1, 1)))
  } else {
    glue::glue("skipping gene {gene} because it was not detected")
  }
}

genes <- c("Homo-w-AAV-GLUL", "Homo-w-AAV-CYP2E1", "Homo-w-AAV-ASS1", "Homo-w-AAV-ALB", "Homo-w-AAV-CYP2F2", "Homo-w-AAV-CYP1A2", "Homo-w-AAV-ARG1", "Homo-w-AAV-PCK1", "Homo-w-AAV-ASL", "Homo-w-AAV-GSTM3", "Homo-w-AAV-SDHD")
genes_of_interest(genes)

```


```{r zonationGenesSamples, fig.width=12, fig.height=8}


for (gene in genes) {
  gene_of_interest(gene)
}

```


### Genes of interest - proliferation

```{r ProliferationGenesCombined}
genes <- c("Homo-w-AAV-LGR5", "Homo-w-AAV-AXIN2")
genes_of_interest(genes)

```

```{r proliferationGenesIndividual}
for (gene in genes) {
  gene_of_interest(gene)
}

```

### Genes of interest - transduction



```{r transductionGenesCombined}
genes <- c("Homo-w-AAV-KIAA0319L", "Homo-w-AAV-GPR108")
genes_of_interest(genes)

```


```{r transductionGenesIndividual}
for (gene in genes) {
  gene_of_interest(gene)
}

```


### Genes of interest - response to AAV

```{r AAVResponseGenesCombined}
genes <- c("Homo-w-AAV-TP53", "Homo-w-AAV-TP21", "Homo-w-AAV-IFI16", "Homo-w-AAV-TLR2")
genes_of_interest(genes)

```

```{r AAVResponseGenesIndividual}
for (gene in genes) {
  gene_of_interest(gene)
}

```

### Genes of interest - AAV expression

```{r AAVExprCombined}
aav_genes <- rownames(cells)[stringr::str_detect(rownames(cells), "VENUS|CERULEAN")]
if (length(aav_genes) == 0) {
  print("skipping this analysis because no AAV 'genes' found")
} else {
  genes_of_interest(aav_genes)
}

```

```{r AAVExprIndividual}
for (gene in aav_genes) {
  gene_of_interest(gene)
}

```
```{r AAVExprJoint1}
FeaturePlot(cells, features = aav_genes, 
            order=TRUE, blend=TRUE, cols=c('lightgray', 'red', 'blue'))
```

Coexpression of Venus and Cerulean (all cells from all samples):

```{r AAVExprJoint2}
 FeaturePlot(subset(cells, dataset_type== 'LK03C_REDHV' | dataset_type== 'LK03V_REDHC'), 
             reduction='umap', features = aav_genes, 
            order=TRUE, split.by="dataset_type")
```

For only the LK03C_REDHV samples:

```{r AAVExprJoint_LK03C_REDHV fig.width=7, fig.height=3}
 FeaturePlot(subset(cells, dataset_type== 'LK03C_REDHV'), 
             reduction='umap', features = aav_genes, 
            order=TRUE, blend=TRUE, cols=c('lightgrey','red', 'blue'))

```

For only the LK03V_REDHC samples:

```{r  AAVExprJoint_LK03V_REDHC, fig.width=7, fig.height=3}
 FeaturePlot(subset(cells, dataset_type== 'LK03V_REDHC'), 
             reduction='umap', features = aav_genes, 
            order=TRUE, blend=TRUE, cols=c('lightgrey','red', 'blue'))

```


## Pseudotime

```{r pseudotime}
library(SeuratWrappers)
library(monocle3)
options(stringsAsFactors = FALSE)

cells.monocle <- as.cell_data_set(cells)
cells.monocle <- cluster_cells(cells.monocle)

p1 <- plot_cells(cells.monocle, show_trajectory_graph = FALSE)
p2 <- plot_cells(cells.monocle, color_cells_by = "partition", show_trajectory_graph = FALSE)

cowplot::plot_grid(p1, p2)

```


## Session info

```{r}
sessionInfo()
```